packages/lesswrong/server/callbacks/rateLimits.ts

```diff
@@ -81,6 +234,105 @@ async function enforcePostRateLimit (user: DbUser) {
   }
 }
 
+
+export async function rateLimitDateWhenUserNextAbleToPost(user: DbUser): Promise<RateLimitInfo|null> {
+  // Admins and Sunshines aren't rate-limited
+  if (shouldIgnorePostRateLimit(user)) return null;
+  
+  // does the user have a moderator-assigned rate limit?
+  const modRateLimitHours = await getModRateLimitHours(user._id);
+
+  // what's the longest rate limit timeframe being evaluated?
+  const maxPostAutolimitHours = getMaxAutoLimitHours(forumSelect(autoRateLimits).posts)
+  const maxHours = Math.max(modRateLimitHours, maxPostAutolimitHours);
+
+  // fetch the posts from within the maxTimeframe
+  const postsInTimeframe = await getPostsInTimeframe(user, maxHours);
+
+  return getStrictestPostRateLimitInfo(user, postsInTimeframe, modRateLimitHours);
+}
+
+export async function rateLimitDateWhenUserNextAbleToComment(user: DbUser, postId: string | null): Promise<RateLimitInfo|null> {
+  const ignoreRateLimits = await shouldIgnoreCommentRateLimit(user, postId);
+  if (ignoreRateLimits) return null;
+
+  // does the user have a moderator-assigned rate limit?
+  const [modRateLimitHours, modPostSpecificRateLimitHours] = await Promise.all([
+    getModRateLimitHours(user._id),
+    getModPostSpecificRateLimitHours(user._id)
+  ]);
+
+  // what's the longest rate limit timeframe being evaluated?
+  const maxCommentAutolimitHours = getMaxAutoLimitHours(forumSelect(autoRateLimits).comments)
+  const maxHours = Math.max(modRateLimitHours, modPostSpecificRateLimitHours, maxCommentAutolimitHours);
+  // fetch the comments from within the maxTimeframe
+  const commentsInTimeframe = await getCommentsInTimeframe(user._id, maxHours);
+
+  return await getStrictestCommentRateLimitInfo({
+    commentsInTimeframe, 
+    user, 
+    modRateLimitHours, 
+    modPostSpecificRateLimitHours, 
+    postId
+  });
+}
+
+function getStrictestRateLimitInfo (rateLimits: Array<RateLimitInfo|null>): RateLimitInfo | null {
+  const nonNullRateLimits = rateLimits.filter(rateLimit => rateLimit !== null) as Array<RateLimitInfo>;
+  const sortedRateLimits = nonNullRateLimits.sort((a, b) => {
+    if (a.nextEligible < b.nextEligible) return 1;
+    if (a.nextEligible > b.nextEligible) return -1;
+    return 0;
+  });
+  return sortedRateLimits[0] || null;
+}
+
+function getStrictestPostRateLimitInfo(user: DbUser, postsInTimeframe: Array<DbPost>, modRateLimitHours: number): RateLimitInfo|null {
+  // for each rate limit, get the next date that user could post  
+  const modRateLimitInfo = getModRateLimitInfo(postsInTimeframe, modRateLimitHours, 1)
+  const autoRatelimits = forumSelect(autoRateLimits).posts
+
+  const autoRateLimitInfos = autoRatelimits ? autoRatelimits?.map(
+    rateLimit => getAutoRateLimitInfo(user, rateLimit, postsInTimeframe)
+  ) : []
```

Comment:
can be simplified to `const autoRateLimitInfos = autoRatelimits?.map(...) ?? []`